### PR TITLE
xdsl: parse custom memref format

### PR DESCRIPTION
**Pull request by superlopuh**
Currently adds memref as if it was a builtin, which feels a little wrong. Would love some input on how to make it better. Should we add custom formats? Is builtin style ok for now?

The number of tests successfully parsed went from 90 to 97 with current changes.

### CODE REVIEW

**Comment by Dinistro** [comment]
> Currently adds memref as if it was a builtin, which feels a little wrong

What do you mean by that? MLIR's `MemRefType` is part of builtin, not the memref dialect: https://mlir.llvm.org/docs/Dialects/Builtin/

File: tests/printer_test.py [hunk 1]
```diff
@@ -633,7 +633,7 @@ def test_parse_dense_mlir():
     """
 
     expected = """
-    %0 = "arith.constant"() {"value" = dense<[[0.0, 1.0, 2.0], [3.0, 4.0, 5.0]]> : tensor<2x3xf64>} : () -> tensor<2x3xf64>
+    %0 = "arith.constant"() {value = dense<[[0.0, 1.0, 2.0], [3.0, 4.0, 5.0]]> : tensor<2x3xf64>} : () -> tensor<2x3xf64>
     """
```
[the remark this answers is not in the document]
Comment (superlopuh) [reply tests/printer_test.py:637]: It's related to the test, I wanted to add a failing test from the test suite verbatim, and check that the output was the same as the input. I can take this change out.

**Comment by superlopuh** [comment]
Oh I misunderstood, in MLIR it's in a separate `MemRef` dialect

**Comment by Dinistro** [comment]
> Oh I misunderstood, in xDSL it's in a separate `MemRef` dialect

Yeah, I remember that we added it there, but there seems to be no reason to differ from MLIR on that.

File: tests/printer_test.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (math-fehr) [reply tests/printer_test.py:637]: Yes, could we remove that change?
I prefer printing the most generic MLIR code, to avoid corner cases. For instance, here, if we have an identifier that has spaces in it, or numbers at the beginning, this will failed to be parsed again by MLIR.

File: xdsl/parser.py [hunk 2]
```diff
@@ -838,6 +839,17 @@ def parse_optional_mlir_vector(self,
             return VectorType.from_type_and_list(typ, dims)
         return None
 
+    def parse_optional_mlir_memref(self,
```
Comment (math-fehr) [inline xdsl/parser.py:842]: Nice, could you also add the unranked memref? This should be implemented like the `UnrankedTensorType`.

**Review by math-fehr — APPROVED** [review]
Nice! I just added two small comments

**Review by math-fehr — APPROVED** [review]
Perfect!
